### PR TITLE
mixer: disable L2 cache

### PR DESCRIPTION
**Pull request by kyessenov**
Ref: #9596 
/assign @mandarjog 
/assign @geeknoid

### CODE REVIEW

**Comment by mandarjog** [comment]
@kyessenov mixer e2e tests are failing, are there tests that rely on cache being enabled?

**Comment by kyessenov** [comment]
@mandarjog No, I don't think we have e2e tests for the L2 cache. The same change did not cause errors in 1.0 branch. I'll investigate why the tests are failing (assuming it's not just flakes).

/retest

File: install/kubernetes/helm/subcharts/mixer/templates/deployment.yaml [hunk 1]
```diff
@@ -28,6 +28,7 @@
           - --monitoringPort={{ .Values.global.monitoringPort }}
           - --address
           - unix:///sock/mixer.socket
+          - --numCheckCacheEntries=0
```
Comment (douglas-reid) [inline install/kubernetes/helm/subcharts/mixer/templates/deployment.yaml]: should this be made the default value for the arg ?

Comment (douglas-reid) [reply install/kubernetes/helm/subcharts/mixer/templates/deployment.yaml]: https://github.com/istio/istio/blob/df0d817c5ecf6f3fab94cb599ddd90c363c4d533/mixer/pkg/server/args.go#L127

Comment (kyessenov) [reply install/kubernetes/helm/subcharts/mixer/templates/deployment.yaml]: I'm not it's a good idea to _change_ defaults. If someone customized the mixer deployment file already, then it's an extra (and not as obvious) thing to miss. Explicit flag is very obvious.
What do you think?

**Comment by kyessenov** [comment]
Build fails in tests :/

File: install/kubernetes/helm/subcharts/mixer/templates/deployment.yaml (hunk 1 above)
Comment (douglas-reid) [reply install/kubernetes/helm/subcharts/mixer/templates/deployment.yaml]: Generally, I'd agree with you. But if the check cache is *fundamentally* broken, I don't see how we could, in good conscience, leave it enabled by default.

@geeknoid @mandarjog thoughts here?

Comment (kyessenov) [reply install/kubernetes/helm/subcharts/mixer/templates/deployment.yaml]: I found one cause of the problem. It's caused by the change of the underlying type for the stringmap values here:

https://github.com/istio/istio/blob/master/mixer/pkg/checkcache/keyShape.go#L198-200

At some point it got swapped to `attribute.StringMap` and the case statement did not have a default clause. 

I think the lesson here is that we have insufficient test coverage for the check cache. I am not convinced this is the only mistake in the cache.

**Comment by geeknoid** [comment]
If the cache is fundamentally broken, then we should turn it off in the
code rather than config, so that it is guaranteed not to be used.

The StringMap string is an excellent catch. I agree this code needs more
testing. Do you think there are reported failures which aren't explained by
this map bug?

**Comment by kyessenov** [comment]
I suggest we turn it off in code by changing the default size to 0 (as well as setting the flag to 0 to make it obvious).

Yes, there's another non-deterministic error that does not involve string maps. I'm having trouble reproducing that one.

We also need to track down `map[string]string` vs `attribute.StringMap` across the entire code base, and that is going to take some time. There are several

**Comment by kyessenov** [comment]
Superseded by #10260

**Pull request closed without merging**